src/core/mac/mac_frame.cpp

```diff
@@ -896,6 +912,143 @@ const uint8_t *Frame::GetFooter(void) const
     return GetPsdu() + GetPsduLength() - GetFooterLength();
 }
 
+void Frame::GenerateImmAck(const Frame *aFrame, bool aIsFramePending)
+{
+    memset(this->mPsdu, 0, this->mLength);
+
+    this->mPsdu[0] |= kFcfFrameAck;
+    if (aIsFramePending)
+    {
+        this->mPsdu[0] |= kFcfFramePending;
+    }
+    reinterpret_cast<uint16_t *>(this->mPsdu)[0] |= aFrame->GetVersion();
+
+    this->mPsdu[2] = aFrame->GetSequence();
+
+    this->mLength = kImmAckLength;
+}
+
+#if OPENTHREAD_CONFIG_MAC_HEADER_IE_SUPPORT
+void Frame::GenerateEnhAck(const Frame *aFrame, bool aIsFramePending, const uint8_t *aIeData, uint8_t aIeLength)
+{
+    uint8_t footerLength;
+
+    SetFrameControlFieldForAck(aFrame, aIsFramePending, aIeLength, this);
+    this->mPsdu[2] = aFrame->GetSequence();
+    SetDstPanIdAndAddrForAck(aFrame, this);
+    SetSecurityHeaderForAck(aFrame, this);
+
```

Comment:
The methods `SetFrameControlFieldForAck()`,  `SetDstPanIdAndAddrForAck()` and `SetSecurityHeaderForAck` are only used from here now. Will we have any other places to call them in future? If not, how about including the implementation directly here?